tests/lib/rules/curly.js

```diff
@@ -231,6 +231,100 @@ ruleTester.run("curly", rule, {
             // https://github.com/feross/standard/issues/664
             code: "if (true) foo()\n;[1, 2, 3].bar()",
             options: ["multi-line"]
+        },
+
+        // https://github.com/eslint/eslint/issues/11908 (also in invalid)
+        {
+            code: "if (true) { const a = 1; }",
+            options: ["multi"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "if (true) { let a; }",
+            options: ["multi"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "if (true) { class a{} }",
+            options: ["multi"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "if (true) { function a() {} }",
+            options: ["multi"]
+        },
+        "if (true) function a() {}",
+        {
+            code: "if (true) \n function a() {}",
+            options: ["multi-line"]
+        },
+        {
+            code: "if (true) function a() \n {}",
+            options: ["multi-line"]
+        },
+        {
+            code: "if (true) function a() \n {}",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (true) { let a; } else { foo; }",
+            options: ["multi", "consistent"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "if (true) { function a() {} } else { foo; }",
+            options: ["multi", "consistent"]
+        },
+        {
+            code: "if (true) function a() {} else b;",
+            options: ["multi", "consistent"]
+        },
+        {
+            code: "if (true) function a() {} else { foo; bar; }",
+            options: ["multi", "consistent"]
+        },
+        {
+            code: "if (true) function a() {} else { let a }",
+            options: ["multi", "consistent"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "if (true) function a() {} else { function b() {} }",
+            options: ["multi", "consistent"]
```

Comment:
I'm confused. Why is it valid for the consequent section to have no block and the alternate section to have a block?

Is it that this is really "invalid", but the block can't be removed in the alternate section because it would change the scope/hoisting of the function?

Reply:
The idea is to treat both as 'valid', thus keep the scope of this rule to only whether the user should simply put or remove braces around certain statements. Cases like this would need additional steps if not a serious refactoring to fix the code. 

This is in line with how the `multi-or-nest` option is already implemented: a single-line function declaration in a single-statement block is not an error (I'm not able to verify this again  at the moment, though). This PR implements the same for the `multi` option, but also prevents reporting and fixing function declarations that are not in a block regardless of the option, so now there are situations where a chain cannot be consistent.

Also, this 'wrong' usage of function declarations will be reported by `no-inner-declarations` (included in `eslint:recommended`).

Reply:
This is how the rule already works with `multi-or-nest` option, which was fixed in #11675:

```js
/*eslint curly: ["error", "multi-or-nest"]*/

// warning
if (foo) {    
    baz();
}

// no warning
if (foo) {   
   function x() {}
}
```

The second block would be 'invalid' by this option because it's one statement in one line, but since it has a lexical declaration (the block cannot be removed) it is treated as 'valid' ( = no warning). I think it's okay to not report this block.

When the `consistent` modifier is on, this block will even force braces around other one-liners in the same chain:

```js
/*eslint curly: ["error", "multi-or-nest", "consistent"]*/

// warning on `if`, not on `else`
if (foo)  
    baz();
else {
   function x() {}
}
```

auto-fixed to:

```js
/*eslint curly: ["error", "multi-or-nest", "consistent"]*/

if (foo) 
    {baz();}
else {
   function x() {}
}
```

This PR does the same with the `multi` option (which wasn't fixed earlier), and the rule will never report a block that has a lexical declaration inside.

In addition, this PR also fixes the behavior of this rule in cases when a function declaration is not in a block. The presumption is that these two statements are not equivalent:

```js
if (foo) 
   function x() {}

if (foo) 
  {function x() {}}
```

so, to be consistent with the fact that 'unchangeable' statements shouldn't be reported, after this PR the rule will not report this:

```js
/*eslint curly: ["error"]*/
// default option, blocks are always required

// no warning
if (foo) 
   function x() {}
```

currently, which I believe is not correct, the rule reports and fixes this code to:

```js
/*eslint curly: ["error"]*/
// default option, blocks are always required

if (foo) 
   {function x() {}}
```

The problem with the `consistent` modifier is the presumption that all statements can be safely surrounded by braces, which leads to situations like this:

```js
/*eslint curly: ["error", "multi", "consistent"]*/

if (foo) 
   function x() {}
else {
    function x() {}
}
```

This isn't a `consistent` chain and it can't be, but I think it's okay to leave it as is ( = not report anything).